vpnglean/CMakeLists.txt

```diff
@@ -48,73 +73,81 @@ if(NOT ${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten")
-    # See: https://github.com/bincrafters/community/issues/466#issuecomment-426329142
-    #      and https://gitlab.cern.ch/-/snippets/2128
-    # 2. Cmake is simply ignoring or being weird about adding these files as
-    # sources to the main target and marking them as GENERATED TRUE.
-    # See: https://gitlab.kitware.com/cmake/cmake/-/issues/18399
-    #
-    # The solution is to generate empty files at build setup time so that we can
-    # add them as sources to other target and still not have Cmake freak out.
-    # Actual generation happen when the `run_glean_parser` command is executed.
-    file(WRITE ${CMAKE_SOURCE_DIR}/src/apps/vpn/glean/generated/metrics.h)
-    file(WRITE ${CMAKE_SOURCE_DIR}/src/apps/vpn/glean/generated/pings.h)
+if(GLEAN_PARSER_VERSION VERSION_GREATER_EQUAL 5.0)
+    # If the glean parser exists, then we can build the telemetry bindings.
+    set(GENERATE_GLEAN_CMD
+        ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/glean_parser_ext/run_glean_parser.py
+                --outdir ${CMAKE_CURRENT_BINARY_DIR}/glean/generated
+    )
+elseif(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/prebuilt)
+    # On some build platforms, we may not have the ability to install the 
+    # glean_parser module, and we must rely on a pre-build script to generate
+    # the telemetry bindings for us.
+    set(GENERATE_GLEAN_CMD
+        ${CMAKE_COMMAND} -E copy_directory
+            ${CMAKE_CURRENT_SOURCE_DIR}/prebuilt/glean/generated
+            ${CMAKE_CURRENT_BINARY_DIR}/glean/generated
+    )
 else()
-    set(GENERATE_GLEAN_CMD echo "Skipping glean_parser step.")
+    message(FATAL_ERROR
+        "Glean generated files are missing, and the glean_parser was not found.\n"
+        "Please run: python3 ./vpnglean/glean_parser_ext/run_glean_parser.py"
+    )
 endif()
 
 ## Generate metrics and pings files
 add_custom_command(
     OUTPUT
-        ${CMAKE_SOURCE_DIR}/src/apps/vpn/glean/generated/metrics.h
-        ${CMAKE_SOURCE_DIR}/src/apps/vpn/glean/generated/pings.h
-        ${CMAKE_CURRENT_SOURCE_DIR}/src/generated/metrics.rs
-        ${CMAKE_CURRENT_SOURCE_DIR}/src/generated/pings.rs
+        ${CMAKE_CURRENT_BINARY_DIR}/glean/generated/metrics.cpp
+        ${CMAKE_CURRENT_BINARY_DIR}/glean/generated/metrics.h
+        ${CMAKE_CURRENT_BINARY_DIR}/glean/generated/pings.cpp
+        ${CMAKE_CURRENT_BINARY_DIR}/glean/generated/pings.h
+        ${CMAKE_CURRENT_BINARY_DIR}/glean/generated/metrics.rs
+        ${CMAKE_CURRENT_BINARY_DIR}/glean/generated/pings.rs
     DEPENDS
         ${CMAKE_CURRENT_SOURCE_DIR}/glean_parser_ext/run_glean_parser.py
+        ${CMAKE_CURRENT_SOURCE_DIR}/glean_parser_ext/templates/cpp_metrics_header.jinja2
+        ${CMAKE_CURRENT_SOURCE_DIR}/glean_parser_ext/templates/cpp_metrics_source.jinja2
+        ${CMAKE_CURRENT_SOURCE_DIR}/glean_parser_ext/templates/cpp_pings_header.jinja2
+        ${CMAKE_CURRENT_SOURCE_DIR}/glean_parser_ext/templates/cpp_pings_source.jinja2
```

Comment:
Hehe, you got me. I was too lazy to split that up into header and cpp file 😳

Reply:
It isn't really necessary, I just think it's a little cleaner this way.

Reply:
Yeah I know and I agree it's cleaner.